aten/src/ATen/native/mps/operations/Indexing.mm

```diff
@@ -28,6 +29,136 @@
+      id<MTLArgumentEncoder> argumentEncoder = [[indexKernelFunction newArgumentEncoderWithBufferIndex:0] autorelease];
+      NSUInteger argumentBufferLength = argumentEncoder.encodedLength;
+      id<MTLBuffer> indexAB = [[device newBufferWithLength:argumentBufferLength options:0] autorelease];
+      [argumentEncoder setArgumentBuffer:indexAB offset:0];
+
+      for (uint32_t idx = 0; idx < num_indices; idx++) {
+        const Tensor& indexTensor = iter.tensor(idx+2);
+        [argumentEncoder setBuffer: getMTLBufferStorage(indexTensor)
+                            offset: indexTensor.storage_offset() * indexTensor.element_size()
+                           atIndex: idx];
+        TORCH_CHECK(indexTensor.scalar_type() == ScalarType::Long, "index(): Expected dtype int64 for Index");
+      }
+
+      // FIXME: PSO needs to be cached
+      id<MTLComputePipelineState> indexSelectPSO = [[device newComputePipelineStateWithFunction: indexKernelFunction
+                                                                                          error: &error] autorelease];
+      TORCH_CHECK(indexSelectPSO, "Failed to created pipeline state object, error: ", [[error description] UTF8String]);
+
+      for (uint32_t idx = 0; idx < num_indices; idx++) {
+        const Tensor& indexTensor = iter.tensor(idx+2);
+        [computeEncoder useResource:getMTLBufferStorage(indexTensor) usage:MTLResourceUsageRead];
+      }
+
+      [computeEncoder setComputePipelineState:indexSelectPSO];
+      [computeEncoder setBuffer:indexAB offset:0 atIndex:0];
+      [computeEncoder setBytes:index_size.data() length:sizeof(index_size[0]) * index_size.size() atIndex:1];
+      [computeEncoder setBytes:index_stride.data() length:sizeof(index_stride[0]) * index_stride.size() atIndex:2];
+      [computeEncoder setBuffer:kernelDataOffsets offset:0 atIndex:3];
+      [computeEncoder setBuffer:inputBuffer offset:inputTensor.storage_offset() * inputTensor.element_size() atIndex:4];
+      [computeEncoder setBuffer:outputBuffer offset:outputTensor.storage_offset() * outputTensor.element_size() atIndex:5];
+
+      NSUInteger tgSize = indexSelectPSO.maxTotalThreadsPerThreadgroup;
+      if (tgSize > numThreads)
+          tgSize = numThreads;
+
+      MTLSize threadGroupSize = MTLSizeMake(tgSize, 1, 1);
+      [computeEncoder dispatchThreads: gridSize
+                threadsPerThreadgroup: threadGroupSize];
+
+      [computeEncoder endEncoding];
+      mpsStream->commit(true);
+    }
+  });
+
+  return true;
+}
+
+void index_kernel_mps(TensorIteratorBase& iter, IntArrayRef index_size, IntArrayRef index_stride) {
+  using namespace mps;
+
+  @autoreleasepool {
+    int64_t num_indices = index_size.size();
+
+    AT_ASSERT(num_indices == index_stride.size());
+    AT_ASSERT(num_indices == iter.ntensors() - 2);
+    const Tensor& inputTensor = iter.tensor(1);
+
+    TORCH_CHECK(c10::isIntegralType(inputTensor.scalar_type(), /*includesBool=*/true), 
```

Comment:
This would return false for `Float` and `half`, which I believe is not an intended behavior.(Also, is it covered by tests right now? If not, then imo it should be)
```suggestion
    TORCH_CHECK(inputTensor.scalar_type() == ScalarType::Float ||
                                 inputTensor.scalar_type() == ScalarType::Half ||
                                c10::isIntegralType(inputTensor.scalar_type(), /*includesBool=*/true), 
```

Reply:
Yes, float and half are covered in the test cases (latest commit should contain this change):
```
   TORCH_CHECK(c10::isIntegralType(inputTensor.scalar_type(), /*includesBool=*/true) ||
                inputTensor.scalar_type() == ScalarType::Float ||
                inputTensor.scalar_type() == ScalarType::Half,
```